File: OSLC-TFS-Consumer/TFSServerEventHandler/TFSServerEventHandler/Mapping/ProductMapper.cs
/*
* Copyright (C) 2015 Ericsson AB. All rights reserved.
*
* Redistribution and use in source and binary forms, with or without
* modification, are permitted provided that the following conditions
* are met:
*
* 1. Redistributions of source code must retain the above copyright
*    notice, this list of conditions and the following disclaimer.
* 2. Redistributions in binary form must reproduce the above copyright
*    notice, this list of conditions and the following disclaimer
*    in the documentation and/or other materials provided with the
*    distribution.
*
* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
* "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
* LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
* A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
* OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
* SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
* LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
* DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
* THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
* (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
* OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

using Microsoft.TeamFoundation.WorkItemTracking.Client;
using Microsoft.TeamFoundation.WorkItemTracking.Server;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Xml.Linq;
using TFSServerEventHandler.Mapping;

namespace TFSServerEventHandler
{
    public class ProductMapper
    {
        private String serviceProvider;

        private Dictionary<String, PRIMProduct> releaseToProductMap;
        private static ProductMapper instance;

        // Provide a list of info/errors that can be shown to user in the TR State field
        public enum EventType
        {
            CreateUpdate,      // Event to be handled
            Disconnect,        // Disconnect Bug
            Ignore,            // Ignore
        }

        public static ProductMapper getInstance()
        {
            if (instance == null)
            {
                instance = new ProductMapper();
            }
            return instance;
        }

        private ProductMapper()
        {        
        }

        public bool Load(String fileName)
        {
            releaseToProductMap = new Dictionary<String, PRIMProduct>();     

            try
            {
                XDocument doc = XDocument.Load(fileName);
                XElement mapping = doc.Element("mapping");

                // Loop over all entries and add to the dictionary 
                foreach (XElement spEntry in mapping.Elements("serviceProvider"))
                {
                    // If the serviceProvider (TFS Project) is defined we only 
                    // care about events that are passed to that TFS Project
                    serviceProvider = spEntry.Attribute("name").Value;

                    foreach (XElement productEntry in spEntry.Elements("product"))
                    {
                        // TODO: Check if product info is case-sensitive
                        String primProdNo = productEntry.Attribute("primProdNo").Value;
                        String primRState = productEntry.Attribute("primRState").Value;
                        PRIMProduct product = new PRIMProduct(primProdNo, primRState);

                        String mapTo = productEntry.Attribute("mapTo").Value; 

                        // NOTE: We don't care about the team attribute for now. 
                        // String team = productEntry.Attribute("team").Value; //.ToLower();;

                        releaseToProductMap.Add(mapTo, product);
                    }
                }
            }
            catch (Exception ex)
            {
                // Error when reading mapping file - assume fatal
                HandlerSettings.LogMessage(
                    "Error when reading the product mapping file: " + fileName +
                    "\nError: " + ex.Message,
                    HandlerSettings.LoggingLevel.ERROR);
                return false;
            }

            return true;
        }

        /// <summary>
        /// The Service Provider / Project for which the mapping applies
        /// </summary>
        public string ServiceProvider { get { return serviceProvider; } }

        /// <summary>
        /// Get the Product based on a Release
        /// </summary>
        public PRIMProduct GetProduct(String release)
        {
            if (releaseToProductMap.ContainsKey(release))
            {
                return releaseToProductMap[release];
            }
            else
            {
                return null;
            } 
        }

        // TODO: This is cheating as we have hardcoded knowledge where release field is.
        // Need to make this fast, but to be correct we should use regular TFSMapper way.
        public bool IsProductInMaintenance(Microsoft.TeamFoundation.WorkItemTracking.Client.WorkItem wi)
        {
            String releaseName = wi.Fields[TFSMapper.TFS_FAULTY_PRODUCT].Value.ToString();
            if (releaseName == null || releaseName.Length == 0)
            {
                // Should not happen - Mapping needs to be initialized
                HandlerSettings.LogMessage(
                    "Can not tell if product is in maintenance as the Release name not defined.",
                    HandlerSettings.LoggingLevel.WARN);
                return false;
            }

            return releaseToProductMap.ContainsKey(releaseName);
        }

        // Check the Release is or just changed from a maintenance product.  
        // TODO: This is cheating as we have hardcoded knowledge where release field is.
        // Need to make this fast, but to be correct we should use regular TFSMapper way.
        public EventType GetEventType(
            Microsoft.TeamFoundation.WorkItemTracking.Client.WorkItem wi,
            WorkItemChangedEvent notification,
            String user)
        {
            Field createTR = wi.Fields[TFSMapper.ERICSSON_DEFECT_CREATETR];
            if (createTR == null ||
               createTR.Value.ToString().Length == 0 ||
               createTR.Value.ToString().Equals("No", StringComparison.OrdinalIgnoreCase))
            {
                return EventType.Ignore;
            }


            // Check project name. If not as defined for product mapping, disregard event.
            // This allows other projects on TFS server to be non-connected to TR system.
            if (serviceProvider.Length > 0 &&
                !wi.Project.Name.Equals(serviceProvider, StringComparison.OrdinalIgnoreCase))
            {
                return EventType.Ignore;
            }

            

            bool inMaint = IsProductInMaintenance(wi);
            if (inMaint)
            {
                // Currenty selected release / product is in maintenance
                EventType evType = CheckDisconnect(wi, notification, user);
                if (evType == EventType.Disconnect)
                { 
                    return EventType.Disconnect;
                }
                return EventType.CreateUpdate;
            }

            // ========================================================================
            // Check if we just changed from having a product in maintenance. Rare case,
            // then need to allow severing links and updating state of the workitem.

            if (notification.ChangeType == ChangeTypes.New)
            {
                // Can only happen for a changed WI, not a new.
                return EventType.Ignore;
            }

            if (notification.ChangedFields == null || notification.ChangedFields.StringFields == null)
            {
                // No fields of interest are changed
                return EventType.Ignore;
            }

            StringField[] changedFields = notification.ChangedFields.StringFields;
            for (int i = 0; i < changedFields.Length; i++)
            {
                String name = changedFields[i].ReferenceName;

                // If the release is changed, we need to update the TR product
                // Note: In case we also set to Duplicate, we will disregard product change
                if (name.Equals(TFSMapper.TFS_FAULTY_PRODUCT))
                {
                    String releaseName = changedFields[i].OldValue;
                    if (releaseName == null || releaseName.Length == 0)
                    {
                        // Should not happen - Mapping needs to be initialized
                        HandlerSettings.LogMessage(
                            "Can not tell if product is in maintenance as the Release name not defined.",
                            HandlerSettings.LoggingLevel.WARN);
                        return EventType.Ignore;
                    }

                    if (releaseToProductMap.ContainsKey(releaseName))
                    {
                        return EventType.CreateUpdate;
                    }
                    else
                    {
                        return EventType.Ignore;
                    }
                }
            }

            return EventType.Ignore;
        }

        // Check if we should disconnect a TR. Condition is that an external save
        // (by TFSProviderUser) has caused the TR Link field to transition from a 
        // value to an empty value, This indicates disconnect.
        private EventType CheckDisconnect(
            Microsoft.TeamFoundation.WorkItemTracking.Client.WorkItem wi,
            WorkItemChangedEvent notification,
            String user)
        {
            if (!user.Equals(HandlerSettings.TFSProviderUser, StringComparison.OrdinalIgnoreCase))
            {
                // Only disconnect based on incoming external event. 
                // Disconnect by user is handled by separate code
                return EventType.Ignore;
            }

            if (notification.ChangeType == ChangeTypes.New)
            {
                // Disconnect can only happen for a changed WI, not a new.
                return EventType.Ignore;
            }

            // Get the current field - Disconnect case assume this is ""
            if (!wi.Fields.Contains(TFSMapper.ERICSSON_DEFECT_LINK_FIELD))
            {
                // Odd case where the mandatory field is missing
                return EventType.Ignore;
            }

            Object trLinkValue = wi.Fields[TFSMapper.ERICSSON_DEFECT_LINK_FIELD].Value;
            if (trLinkValue == null)
            {
                return EventType.Ignore;
            }
            String trLink = trLinkValue.ToString();
            if (trLink != null && trLink.Length > 0)
            {
                return EventType.Ignore;
            }

            TextField[] changedTextFields = notification.TextFields;
            if (changedTextFields == null || changedTextFields.Length == 0)
            {
                // The TR Link field is a Text field - if no changes, ignore.
                return EventType.Ignore;
            }

            // Check so also the TR Link field is changed
            for (int i = 0; i < changedTextFields.Length; i++)
            {
                String name = changedTextFields[i].ReferenceName;
                if (name.Equals(TFSMapper.ERICSSON_DEFECT_LINK_FIELD))
                {
                    return EventType.Disconnect;
                }
            }

            return EventType.Ignore;           
        }    
    }
}

File: OSLC-TFS-Adapter/com.ericsson.eif.tfs.oslc/src/com/ericsson/eif/tfs/oslc/mapping/ProductMapper.java
package com.ericsson.eif.tfs.oslc.mapping;

/*
* Copyright (C) 2015 Ericsson AB. All rights reserved.
*
* Redistribution and use in source and binary forms, with or without
* modification, are permitted provided that the following conditions
* are met:
*
* 1. Redistributions of source code must retain the above copyright
*    notice, this list of conditions and the following disclaimer.
* 2. Redistributions in binary form must reproduce the above copyright
*    notice, this list of conditions and the following disclaimer
*    in the documentation and/or other materials provided with the
*    distribution.
*
* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
* "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
* LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
* A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
* OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
* SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
* LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
* DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
* THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
* (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
* OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;
import java.util.Map.Entry;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import javax.xml.xpath.XPath;
import javax.xml.xpath.XPathConstants;
import javax.xml.xpath.XPathExpressionException;
import javax.xml.xpath.XPathFactory;

import org.w3c.dom.Document;
import org.w3c.dom.NamedNodeMap;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import org.xml.sax.SAXException;

import com.ericsson.eif.tfs.oslc.TFSAdapterManager;

public class ProductMapper {
    
    //TODO - Add support for getting project name from serviceProvider
    //Eg: serviceProviderMap containing a map of projects containing primproduct maps
    // project1
    //    { 
    //      primProductA1,
    //      primProductA2,
    //      primProductA3
    //    }
    // project2
    //    { 
    //      primProductB1,
    //      primProductB2,
    //      primProductB3
    //    }
    

    public class ServiceProvider {
        private String name; // E.g. project in TFS
        private String entity; // E.g. collectionName in TFS

        public ServiceProvider(String name, String entity) {
            this.name = name;
            this.entity = entity;
        }
        
        public String getName() {
            return name;
        }
        
        public String getEntity() {
            return entity;
        }
    }

    ServiceProvider serviceProvider;
    Map<String, Object> serviceProviderMap; //TODO - see above
    Map<PrimProduct, BackendEntity> map;
    private static ProductMapper instance;
    private String productMappingFile;

    private ProductMapper() {
        map = new HashMap<>();
    }

    public static ProductMapper getInstance() {
        if (instance == null) {
            instance = new ProductMapper();
        }
        return instance;
    }

    public void setTestMappingFile(String file) {
        productMappingFile = file;
    }

    public void load(InputStream input) throws ParserConfigurationException,
            SAXException, IOException, XPathExpressionException {
        DocumentBuilderFactory domFactory = DocumentBuilderFactory
                .newInstance();
        domFactory.setNamespaceAware(true);
        DocumentBuilder builder = domFactory.newDocumentBuilder();
        Document doc = builder.parse(input);
        XPath xpath = XPathFactory.newInstance().newXPath();

        // locate all product serviceProvider nodes
        NodeList spNodes = (NodeList) xpath.evaluate(
                "//mapping//serviceProvider", doc, XPathConstants.NODESET);

        for (int i = 0; i < spNodes.getLength(); i++) {
            Node node = spNodes.item(i);
            if (processThis(node)) {
                processServiceProvider(node);
            }
        }
    }

    public void load() {
    	if (productMappingFile == null) {
            productMappingFile = TFSAdapterManager.getProductMappingFile();
    	}
    	
        try {
            File inputFile = new File(productMappingFile);
            if (!inputFile.exists()) {
                // try relative path
                inputFile = new File(TFSAdapterManager.getAdapterServletHome()
                        + File.separator + productMappingFile);
                if (!inputFile.exists()) {
                    TFSAdapterManager
                            .logAndExit("Products mapping file missing: "
                                    + productMappingFile);
                }
            }
            load(new FileInputStream(inputFile));
        } catch (XPathExpressionException | ParserConfigurationException
                | SAXException | IOException e) {
            e.printStackTrace();
        }
    }

    private void processServiceProvider(Node node) {
        NamedNodeMap attributes = node.getAttributes();
        String name = null;
        String collectionName = null;
        for (int i = 0; i < attributes.getLength(); i++) {
            Node attribute = attributes.item(i);
            switch (attribute.getNodeName()) {
            case "name":
                name = attribute.getNodeValue();
                break;
            case "collectionName":
                collectionName = attribute.getNodeValue();
                break;
            default:
                break;
            }
        }
        serviceProvider = new ServiceProvider(name, collectionName);
        NodeList children = node.getChildNodes();
        for (int j = 0; j < children.getLength(); j++) {
            Node child = children.item(j);
            if (processThis(child)) {
                processPrimProduct(child);
            }
        }
    }

    private void processPrimProduct(Node node) {
        NamedNodeMap attributes = node.getAttributes();
        String primProdNo = null;
        String primRState = null;
        String mapTo = null;
        String team = null;
        for (int i = 0; i < attributes.getLength(); i++) {
            Node attribute = attributes.item(i);
            switch (attribute.getNodeName()) {
            case "primProdNo":
                primProdNo = attribute.getNodeValue();
                break;
            case "primRState":
                primRState = attribute.getNodeValue();
                break;
            case "mapTo":
                mapTo = attribute.getNodeValue();
                break;
            case "team":
                team = attribute.getNodeValue();
                break;
            default:
                break;
            }
        }
        PrimProduct product = new PrimProduct(primProdNo, primRState);
        BackendEntity entity = new BackendEntity(mapTo, team);
        map.put(product, entity);
    }

    private boolean processThis(Node node) {
        return node.getAttributes() != null; //
    }
    
    public Iterator<PrimProduct> getProductIterator() {
        return map.keySet().iterator();
    }
    
    public Iterator<Entry<PrimProduct, BackendEntity>> getEntityIterator() {
        return map.entrySet().iterator();
    }

    public BackendEntity getEntity(PrimProduct product) {
        if (map.containsKey(product)) {
            return map.get(product);
        }
        return null;
    }

    public PrimProduct getProduct(BackendEntity entity) {
        if (map.containsValue(entity)) {
            Iterator<Entry<PrimProduct, BackendEntity>> iter = map.entrySet().iterator();
            while (iter.hasNext()) {
                Entry<PrimProduct, BackendEntity> next = iter.next();
                if (next.getValue().equals(entity)) {
                    return next.getKey();
                }
            }
        }
        return null;
    }
    
    public ServiceProvider getServiceProvider() {
        return serviceProvider;
    }
}
